packages/jest-runtime/src/index.js

```diff
@@ -213,7 +213,7 @@ class Runtime {
       platforms: config.haste.platforms || ['ios', 'android'],
       providesModuleNodeModules: config.haste.providesModuleNodeModules,
       resetCache: options && options.resetCache,
-      retainAllFiles: false,
+      retainAllFiles: config.haste.retainAllFiles,
```

Comment:
You need to define `retainAllFiles` as a optional member of `HasteConfig` in `types/Config` and fallback to false:
```js
retainAllFiles: config.haste.retainAllFiles || false
```